File: assignments/caesar_cipher/java/CaesarCipherTestRunner.java
public class CaesarCipherTestRunner {
    public static void main(String[] args) {
        CaesarCipher cipher = new CaesarCipher();

        // Test Cases: [plainText, key, expectedCipherText]
        Object[][] testCases = {
            {"Hello, World!", 1, "Ifmmp, Xpsme!"},
            {"xyz", 3, "abc"},
            {"XYZ", 3, "ABC"},
            {"Be sure to drink your Ovaltine!", 13, "Or fher gb qevax lbhe Binygvar!"},
            {"Ciphering.", 26, "Ciphering."},
            {"middle-Outz", 2, "okffng-Qwvb"},
            {"A", 0, "A"},
            {"Z", 1, "A"},
            {"a", 25, "z"},
            {"123 ABC xyz .,!", 5, "123 FGH cde .,!"},
            {"", 10, ""}, // Empty string test
            {"Test", 52, "Test"} // Key is multiple of 26
        };

        int passedCount = 0;
        StringBuilder report = new StringBuilder(); // StringBuilder is OK for Test Runner output

        System.out.println("===== Starting Automated Tests for Caesar Cipher (Java) =====");

        for (int i = 0; i < testCases.length; i++) {
            String plainText = (String) testCases[i][0];
            int key = (int) testCases[i][1];
            String expected = (String) testCases[i][2];
            String actual = "Error"; // Default
            String caseIdentifier = "Test Case #" + (i + 1) + ": Input \"" + plainText + "\", Key=" + key;

            try {
                actual = cipher.encodeText(plainText, key);
                report.append(caseIdentifier).append("\n");
                report.append(" -> Expected: \"").append(expected).append("\"\n");
                report.append(" ->      Got: \"").append(actual).append("\"");

                if (expected.equals(actual)) {
                    report.append(" -> ✅ PASS\n\n");
                    passedCount++;
                } else {
                    report.append(" -> ❌ FAIL\n\n");
                }
            } catch (Exception e) {
                 report.append(caseIdentifier).append("\n");
                 report.append(" -> ❌ ERROR during encryption: ").append(e.getClass().getSimpleName()).append("\n\n");
            }
        }

        System.out.println(report.toString());
        System.out.println("===== Test Summary =====");
        System.out.println("Total Tests: " + testCases.length); System.out.println("Passed: " + passedCount); System.out.println("Failed: " + (testCases.length - passedCount)); System.out.println("========================");
        System.out.println("[SUMMARY]:" + passedCount + ":" + testCases.length);
    }
}

File: assignments/caesar_cipher/python/caesar_cipher_test_runner.py
import sys
from caesar_cipher import CaesarCipher # Import the student's class

def run_tests():
    solver = CaesarCipher()
    # Test Cases: (plainText, key, expectedCipherText)
    test_cases = [
        ("Hello, World!", 1, "Ifmmp, Xpsme!"),
        ("xyz", 3, "abc"), ("XYZ", 3, "ABC"),
        ("Be sure to drink your Ovaltine!", 13, "Or fher gb qevax lbhe Binygvar!"),
        ("Ciphering.", 26, "Ciphering."), ("middle-Outz", 2, "okffng-Qwvb"),
        ("A", 0, "A"), ("Z", 1, "A"), ("a", 25, "z"),
        ("123 ABC xyz .,!", 5, "123 FGH cde .,!"), ("", 10, ""),
        ("Test", 52, "Test")
    ]
    passed_count = 0
    report = []
    print("===== Starting Automated Tests for Caesar Cipher (Python) =====")

    for i, (plain_text, key, expected) in enumerate(test_cases):
        actual = "Error"
        case_identifier = f'Test Case #{i + 1}: Input "{plain_text}", Key={key}'
        try:
            actual = solver.encode_text(plain_text, key)
            report.append(case_identifier)
            report.append(f' -> Expected: "{expected}"')
            report.append(f' ->      Got: "{actual}"')

            if actual == expected:
                report.append(" -> ✅ PASS\n")
                passed_count += 1
            else:
                report.append(" -> ❌ FAIL\n")
        except Exception as e:
            report.append(case_identifier)
            report.append(f" -> ❌ ERROR during encryption: {type(e).__name__}\n")

    print("\n".join(report))
    print("===== Test Summary ====="); print(f"Total Tests: {len(test_cases)}"); print(f"Passed: {passed_count}"); print(f"Failed: {len(test_cases) - passed_count}"); print("========================")
    print(f"[SUMMARY]:{passed_count}:{len(test_cases)}")

if __name__ == "__main__":
    sys.path.append('.')
    run_tests()